### PR TITLE
Fixes agent error handling when check definition is invalid. Distingu…

### LINKED ISSUE #3559: Agent silently filters out invalid check definitions
Originally uncovered by @magiconair 

When agents load service definition files, it first filters out any [invalid checks](https://github.com/hashicorp/consul/blob/1acff3533ed3f00d15da8c7e9735b6c2f7c9bfc8/agent/structs/service_definition.go#L31) before attempting to [validate them](https://github.com/hashicorp/consul/blob/828be5771afa6790fd2f8feb43b7e5b64f116454/agent/agent.go#L1470). There are no error messages logged and the agent starts up due to this. 

We will change this behavior in 1.0 by making sure the agent fails starting up if the check definition is invalid.

### PR DESCRIPTION
**Pull request by preetapan**
…ishes between empty checks vs invalid checks

This fixes #3559

### CODE REVIEW

File: agent/structs/check_type.go [hunk 1]
```diff
@@ -48,32 +48,47 @@ func (c *CheckType) Valid() bool {
...
+
 // IsScript checks if this is a check that execs some kind of script.
 func (c *CheckType) IsScript() bool {
 	return c.Script != "" || len(c.ScriptArgs) > 0
 }
 
 // IsTTL checks if this is a TTL type
 func (c *CheckType) IsTTL() bool {
-	return c.TTL != 0
+	return c.TTL > 0
```
Comment (preetapan) [inline agent/structs/check_type.go]: We were allowing negative TTLs and intervals for some reason. I fixed that as part of this change. 

For TTL checks there are no other required fields, so this is the only way to distinguish between valid vs empty (0 is the default value when the struct definition for a service is parsed out of json).

File: agent/structs/check_type.go [hunk 2]
```diff
@@ -48,32 +48,47 @@ func (c *CheckType) Valid() bool {
...
 }
 
 // IsDocker returns true when checking a docker container.
 func (c *CheckType) IsDocker() bool {
-	return c.IsScript() && c.DockerContainerID != "" && c.Interval != 0
+	return c.IsScript() && c.DockerContainerID != "" && c.Interval > 0
+}
+
+// ValidateMsg returns a user friendly error message indicating missing fields
+func (c *CheckType) ValidateMsg() string {
```
Comment (preetapan) [inline agent/structs/check_type.go]: Open to better suggestions here, I basically wanted the agent to print a nicer error message than "Invalid check + the full struct with all the different fields in it.

File: agent/structs/service_definition.go [hunk 3]
```diff
@@ -28,12 +30,17 @@ func (s *ServiceDefinition) NodeService() *NodeService {
 	return ns
 }
 
-func (s *ServiceDefinition) CheckTypes() (checks CheckTypes) {
+func (s *ServiceDefinition) CheckTypes() (checks CheckTypes, err error) {
```
Comment (magiconair) [inline agent/structs/service_definition.go]: `else` is redundant if you flip the condition. You can consider it nit picking but the way I view it is that you first clear all the preconditions one after another without branching and then at the end - when all the pre-conditions are met - you execute the body. This way you can read the code top to bottom. 

The other thing is that `s.Checks = append(s.Checks, &s.Check)` has a side effect which you wouldn't expect when calling this method since it returns an entirely new `[]*CheckType`. I'd remove that and move it where it belongs. 

This also looks too object oriented and the other question is why would we filter empty and invalid checks on query instead of preventing them to be there in the first place? `s.Checks` should only have valid, non-empty `*CheckType` instances IMO.

```go
func (s *ServiceDefinition) CheckTypes() (checks CheckTypes, err error) {
	s.Checks = append(s.Checks, &s.Check)
	for _, check := range s.Checks {
		if check.Empty() {
			continue
		}
		if !check.Valid() {
			return nil, fmt.Errorf("invalid check definition:%v", check.ValidateMsg())
		}
		checks = append(checks, check)
	}
	return
}
```

File: agent/structs/service_definition_test.go [hunk 4]
```diff
@@ -32,10 +35,41 @@ func TestAgentStructs_CheckTypes(t *testing.T) {
...
+	checks, err := svc.CheckTypes()
+	if err != nil {
+		t.Fatalf("Got error:%v", err)
+	}
+
+	if len(checks) != 4 {
+		t.Fatalf("bad: %#v", svc)
+	}
+
+	//Error on invalid interval
```
Comment (magiconair) [inline agent/structs/service_definition_test.go]: space after `//`

File: agent/structs/service_definition_test.go [hunk 5]
```diff
@@ -32,10 +35,41 @@ func TestAgentStructs_CheckTypes(t *testing.T) {
...
+
+	checks, err = svc.CheckTypes()
+	expectedErr := fmt.Errorf("invalid check definition:Interval must be >0")
+	verify.Values(t, "", expectedErr.Error(), err.Error())
+
+	if len(checks) != 0 {
+		t.Fatalf("bad: %#v", svc)
+	}
+
+	//Error on invalid ttl
```
Comment (magiconair) [inline agent/structs/service_definition_test.go]: here as well

File: agent/structs/check_type.go [hunk 6]
```diff
@@ -44,8 +46,18 @@ type CheckType struct {
 type CheckTypes []*CheckType
 
 // Valid checks if the CheckType is valid
-func (c *CheckType) Valid() bool {
-	return c.IsTTL() || c.IsMonitor() || c.IsHTTP() || c.IsTCP() || c.IsDocker()
+func (c *CheckType) Valid() (bool, error) {
+	valid := c.IsTTL() || c.IsMonitor() || c.IsHTTP() || c.IsTCP() || c.IsDocker()
+	var err error
+	if !valid {
+		err = fmt.Errorf(c.validateMsg())
```
Comment (slackpad) [inline agent/structs/check_type.go]: The `validateMsg` helper is odd here and kind of confusing how it works. This seems like it should be more explicit about checking the fields here and produce the errors directly from in here. There are some edge cases where this was sloppy where they can give a TTL and an interval, so that's weird as well. I'd change this to something like

```
func (c *CheckType) Validate() error {
    if c.IsScript() || c.HTTP != "" || c.TCP != "" {
        if c.Interval <= 0 {
            return fmt.Errorf("Interval must be >0 for script, HTTP, or TCP checks")
        }
    
        if c.TTL > 0 {
            return fmt.Errorf("Interval and TTL cannot both be specified")
        }
    } else if c.TTL <= 0 {
        return fmt.Errorf("TTL must be >0 for TTL checks")
    }

    return nil
}
```

File: agent/structs/service_definition.go [hunk 7]
```diff
@@ -28,12 +28,19 @@ func (s *ServiceDefinition) NodeService() *NodeService {
...
 	for _, check := range s.Checks {
-		if check.Valid() {
-			checks = append(checks, check)
+		_, err := check.Valid()
+		if err != nil {
+			return nil, err
 		}
 	}
-	return
+	return s.Checks, nil
```
Comment (slackpad) [inline agent/structs/service_definition.go]: I thought we were going to refactor this to return a fresh slice, but never alter `s.Checks`.

File: agent/agent_endpoint.go [hunk 8]
```diff
@@ -344,7 +344,8 @@ func (s *HTTPServer) AgentRegisterCheck(resp http.ResponseWriter, req *http.Requ
 
 	// Verify the check type.
 	chkType := args.CheckType()
-	if !chkType.Valid() {
+	_, err := chkType.Valid()
+	if err != nil {
 		resp.WriteHeader(http.StatusBadRequest)
 		fmt.Fprint(resp, invalidCheckMessage)
```
Comment (slackpad) [inline agent/agent_endpoint.go]: Now that we have a more detailed error, what do you think about adding the error along with the feedback (here and on line 576)?

File: agent/structs/check_type.go (hunk 6 above)
Comment (magiconair) [reply agent/structs/check_type.go]: I find that a bit hard to follow. How about:

```go
func (c *CheckType) Validate() error {
	intervalCheck := c.IsScript() || c.HTTP != "" || c.TCP != ""

	if c.Interval > 0 && c.TTL > 0 {
		return fmt.Errorf("Interval and TTL cannot both be specified")
	}
	if intervalCheck && c.Interval <= 0 {
		return fmt.Errorf("Interval must be > 0 for Script, HTTP, or TCP checks")
	}
	if !intervalCheck && c.TTL <= 0 {
		return fmt.Errorf("TTL must be > 0 for TTL checks")
	}
	return nil
}
```

File: agent/structs/service_definition.go (hunk 7 above)
Comment (preetapan) [reply agent/structs/service_definition.go]: Maybe I misunderstood, I thought we wanted to stop doing that and maintain s.Checks correctly. Can fix it though.

File: agent/structs/check_type.go (hunk 6 above)
Comment (slackpad) [reply agent/structs/check_type.go]: @magiconair's is cleaner than my suggestion

File: agent/structs/service_definition.go (hunk 7 above)
Comment (slackpad) [reply agent/structs/service_definition.go]: We wanted to eliminate that side-effect.
